app/tasks.py

```diff
@@ -459,3 +464,104 @@ def get(self):
+
+    # Lifetime of a single task is 10 min. It stops fetching and starts
+    # uploading after 4 min, assuming uploading takes similar time as
+    # fetching.
+    MAX_FETCH_TIME = datetime.timedelta(minutes=4)
+    
+    def __init__(self, *args, **kwargs):
+        super(DumpCSV, self).__init__(*args, **kwargs)
+        self.storage = cloud_storage.CloudStorage()
+
+    def task_name(self):
+        return 'dump-csv'
+
+    def schedule_next_task(self, cursor, timestamp):
+        """Schedule the next task for to carry on with this query.
+        """
+        self.add_task_for_repo(
+                self.repo,
+                self.task_name(),
+                self.ACTION,
+                cursor=cursor,
+                timestamp=str(calendar.timegm(timestamp.utctimetuple())))
+
+    def get(self):
+        if self.repo:
+            self.run_task_for_repo(self.repo)
+        else:
+            # Sets lifetime of objects in Google Cloud Storage to 2 days. It is
+            # enough to call this only once for the bucket, but here is just a
+            # convenient place to call it.
+            self.storage.set_objects_lifetime(lifetime_days=2)
+
+            for repo in model.Repo.list():
+                self.add_task_for_repo(repo, self.task_name(), self.ACTION)
+
+    def run_task_for_repo(self, repo):
+        start_time = utils.get_utcnow()
+        timestamp = self.params.timestamp or start_time
+        base_name = '%s-persons-%s' % (
+            repo, timestamp.strftime('%Y-%m-%d-%H%M%S'))
+        is_first = not self.params.cursor
+
+        query = model.Person.all_in_repo(repo).order('entry_date')
+        if self.params.cursor:
+            query.with_cursor(self.params.cursor)
+
+        csv_io = StringIO.StringIO()
+        writer = record_writer.PersonCsvWriter(csv_io, write_header=is_first)
+
+        has_data = False
+        scan_completed = False
+        while True:
+            persons = query.fetch(limit=FETCH_LIMIT)
+            if persons:
+                has_data = True
+            else:
+                scan_completed = True
+                break
+            records = [PFIF.person_to_dict(person) for person in persons]
+            # So far it only supports dump of records without sensitive fields.
+            utils.filter_sensitive_fields(records)
+            writer.write(records)
+            if utils.get_utcnow() >= start_time + self.MAX_FETCH_TIME:
+                break
+            query.with_cursor(query.cursor())
+
+        final_csv_name = '%s.csv' % base_name
+        temp_csv_name = '%s.temp.csv' % base_name
+
+        if is_first:
+            self.storage.insert_object(final_csv_name, 'text/csv', '')
+
+        if has_data:
+            # Creates a temporary CSV file with new records, and append it to
+            # the final CSV file.
+            self.storage.insert_object(
+                temp_csv_name, 'text/csv', csv_io.getvalue())
+            self.storage.compose_objects(
```

Comment:
Do you have a sense of how long Cloud Storage would take to merge these files when it gets very large? Is there any risk of it over-running the 10 minutes the task has?

Reply:
My guess is that it actually doesn't perform concatenation, and instead creates an object with pointers to the source objects internally. Thus it would be O(1), not O(byte_size). I guessed it because of concept of "object counts":
https://cloud.google.com/storage/docs/composite-objects

I will monitor the production job to see if it's actually the case.